### PR TITLE
Unified Config

### PR DESCRIPTION
**Pull request by semperos**
These changes reduce the surface area of the library to one function:

```clj
(rp.config/load-config :ci)
```

You can optionally specify an EDN file other than `config.edn` to load.

This function combines environment and application configuration loading into one step, with all values that the user cares about needing to be specified in the application config EDN.

To get environment variable values into the EDN, these changes support a `#rp.config/env-var` reader literal that allows templating your EDN with references to environment variables. For example, if I wanted to get the value of the `TMPDIR` environment variable in my application config:

```clj
{:tmp-dir #rp.config/env-var "TMPDIR"}
```

When `load-config` reads this in, it will attempt to find `TMPDIR` in the available environment variables. If not found, `nil` is put in its place (open for discussion if folks feel strongly).

In addition, this PR supports the `#rp.config/var` literal which will resolve the namespace-qualified symbol or keyword after it as a Clojure var. This is useful when you need to add references to functions (think web handler or component constructors) in your config.

_N.B.:_ Code has been copied from the clj-infra project for environment loading. If a local file of `.env.ci` for CI, `.env.qa` for QA, etc., is not found, the code will reach out to the Infra server to get the latest values and cache them locally.

### CODE REVIEW

File: src/rp/config.clj [hunk 1]
```diff
@@ -0,0 +1,29 @@
+(ns rp.config
+  (:require [rp.config.app :as app]
+            [rp.config.env :as env]))
+
+(def ^:const default-config-file-name "config.edn")
+
+(def default-reader-opts
+  "Allow enforcing the resolution of var references when Liftoff config is read."
```
Comment (err) [inline src/rp/config.clj]: reference to liftoff

Comment (semperos) [inline src/rp/config.clj]: Ah, you caught me! :)

**Comment by semperos** [comment]
@err No, it doesn't throw exceptions if environment variables aren't found. This code _will_ throw an exception if there isn't a local file for environment variables and, in the absence of that file, if the code can't reach the Infra server to retrieve the environment variable values.

If you or others feel strongly, we can certainly make this code throw an exception and make calling code catch exceptions and provide defaults at that point. I'd personally like to decentralize the concern of validating config to components that need to use it, and with that in mind having `nil` (which is the behavior of this PR) is more conducive to that.

@err Thoughts?

**Comment by err** [comment]
I'm open to seeing how it plays out. It makes sense to me to have a component determine for itself whether its supplied config is sufficient, and throw an exception if not.

There's a slight inconsistency between app and env vars because we still have [this check](https://github.com/rentpath/clj-config/blob/d7ff9efe8542dc838d2fb21881f744326fbde6fd/src/rp/config/reader.clj#L123) for a missing map-entry pair on config.edn values. Though maybe we want to keep that for now, as I suppose it highlights a malformed app-config map. One wonders whether that can be encoded as part of the spec (not a priority, though).

Anyway, thanks for explaining.
 👍  on the PR as a whole

**Comment by semperos** [comment]
Good points, @err , thanks for discussing. I agree with your explanation, that if you do specify legal environments in the app config and then customize a config value based on env but neglect to handle all cases, that represents a malformed app config map. I think that check is orthogonal to whether or not specific environment variables have values.

With that said, as I think about it, the case of providing a fallback for a lack of environment variable has always been to allow having a single code path work for different environments. So for example, we'd never have a `PORT` env var locally, so our code for booting a web server would handle `PORT` not being defined in dev/test but would defer to it for production. That should be handled directly in the app config using different environment specifiers, and only production, in this example, would rely on the `PORT` env var in the config map via `#rp.config/env-var`.

Your hesitation has swayed me :) I will make the code throw an exception if an environment variable is not found.

**Pull request merged**